File: mobile/lib/features/device/domain/device.dart
/// Device domain model
class Device {
  final int? id;
  final String name;
  final String? deviceId;
  final String? description;
  final String? deviceType;
  final String? location;
  final bool? isActive;
  final DateTime? lastSeen;
  final DateTime? createdAt;

  Device({
    this.id,
    required this.name,
    this.deviceId,
    this.description,
    this.deviceType,
    this.location,
    this.isActive,
    this.lastSeen,
    this.createdAt,
  });

  factory Device.fromJson(Map<String, dynamic> json) {
    return Device(
      id: json['id'] as int?,
      name: json['name'] as String,
      deviceId: json['deviceId'] as String?,
      description: json['description'] as String?,
      deviceType: json['deviceType'] as String?,
      location: json['location'] as String?,
      isActive: json['isActive'] as bool?,
      lastSeen: json['lastSeen'] != null
          ? DateTime.parse(json['lastSeen'] as String)
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'name': name,
      if (deviceId != null) 'deviceId': deviceId,
      if (description != null) 'description': description,
      if (deviceType != null) 'deviceType': deviceType,
      if (location != null) 'location': location,
      if (isActive != null) 'isActive': isActive,
      if (lastSeen != null) 'lastSeen': lastSeen!.toIso8601String(),
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
    };
  }

  Device copyWith({
    int? id,
    String? name,
    String? deviceId,
    String? description,
    String? deviceType,
    String? location,
    bool? isActive,
    DateTime? lastSeen,
    DateTime? createdAt,
  }) {
    return Device(
      id: id ?? this.id,
      name: name ?? this.name,
      deviceId: deviceId ?? this.deviceId,
      description: description ?? this.description,
      deviceType: deviceType ?? this.deviceType,
      location: location ?? this.location,
      isActive: isActive ?? this.isActive,
      lastSeen: lastSeen ?? this.lastSeen,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

File: backend/src/main/java/com/aliparwiz/microgreens/device/Device.java
package com.aliparwiz.microgreens.device;

import com.aliparwiz.microgreens.auth.User;
import com.aliparwiz.microgreens.sensor.SensorReading;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Device Entity
 * Represents an IoT device that collects sensor data
 */
@Entity
@Table(name = "devices")
@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(exclude = {"owner", "sensorReadings"})
@ToString(exclude = {"owner", "sensorReadings"})
public class Device {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(nullable = false)
    private String name;
    
    @Column(unique = true)
    private String deviceId; // Unique device identifier (e.g., MAC address, serial number)
    
    private String description;
    
    @Column(name = "device_type")
    private String deviceType; // e.g., "sensor", "controller", "camera"
    
    @Column(name = "location")
    private String location;
    
    @Column(name = "is_active")
    private Boolean isActive = true;
    
    @Column(name = "last_seen")
    private LocalDateTime lastSeen;
    
    @Column(name = "created_at")
    private LocalDateTime createdAt = LocalDateTime.now();
    
    @Column(name = "updated_at")
    private LocalDateTime updatedAt = LocalDateTime.now();
    
    // Many devices belong to one user
    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User owner;
    
    // One device has many sensor readings
    @JsonIgnore
    @OneToMany(mappedBy = "device", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<SensorReading> sensorReadings;
}

